sklearn/utils/validation.py

```diff
@@ -184,6 +184,43 @@ def _is_arraylike(x):
             hasattr(x, '__array__'))
 
 
+def _num_features(X):
+    """Return the number of features in an array-like X.
+
+    Parameters
+    ----------
+    X : array-like
+        array-like to get the number of features.
+
+    Returns
+    -------
+    features : int
+        Number of features
+    """
+    message = "Unable to find the number of features from X"
+    if not hasattr(X, '__len__') and not hasattr(X, 'shape'):
+        if not hasattr(X, '__array__'):
+            raise TypeError(message)
+        X = np.asarray(X)
+
+    if hasattr(X, 'shape') and X.shape is not None:
+        if len(X.shape) <= 1:
```

Comment:
I don't see any legitimate reason to accept and input with `None` as shape or any invalid shape for that matter.

```suggestion
    if hasattr(X, 'shape'):
        if not hasattr(X.shape, '__len__') or len(X.shape) <= 1:
```